src/lib/reasoners/enum_rel.ml

```diff
@@ -28,32 +28,81 @@
 (*                                                                        *)
 (**************************************************************************)
 
-module A  = Xliteral
 module L  = List
-module Hs = Hstring
+module X = Shostak.Combine
+module Ex = Explanation
+module MX = Shostak.MXH
+module SX = Shostak.SXH
+module HSS = Set.Make (Hstring)
+module LR = Uf.LX
+module Th = Shostak.Enum
 
 let timer = Timers.M_Sum
 
-type 'a abstract = 'a Enum.abstract =
-  | Cons of Hs.t * Ty.t
-  | Alien of 'a
+module Domain = struct
+  type t = {
+    constrs : HSS.t;
+    ex : Ex.t;
+  }
 
-module X = Shostak.Combine
-module Sh = Shostak.Enum
+  exception Inconsistent of Ex.t
 
-module Ex = Explanation
+  let[@inline always] cardinal { constrs; _ } = HSS.cardinal constrs
 
-module MX = Shostak.MXH
-module HSS = Set.Make (struct type t=Hs.t let compare = Hs.compare end)
+  let[@inline always] choose { constrs; _ } = HSS.choose constrs
 
-module LR = Uf.LX
+  let domain ~constrs ex =
+    if HSS.is_empty constrs then
+      raise_notrace @@ Inconsistent ex
+    else
+      { constrs; ex }
+
+  let unknown r =
+    match Th.embed r with
+    | Cons (c, _) ->
+      domain ~constrs:(HSS.singleton c) Ex.empty
+    | _ ->
+      match X.type_info r with
+      | Ty.Tsum (_, constrs) ->
+        (* Return the list of all the constructors of the type of [r]. *)
+        let constrs = HSS.of_list constrs in
+        assert (not @@ HSS.is_empty constrs);
+        { constrs; ex = Ex.empty }
+      | _ ->
+        (* Only Enum values can have a domain. This case can happen since we
+           don't dispatch the literals processed in [assume] by their types in
+           the Relation module. *)
+        invalid_arg "unknown"
```

Comment:
Forgot to update, but it looks like this case can no longer happen because we *do* check the types before calling `Domains.add`, so the code is correct but the comment no longer is.

Reply:
So why do you use `invalid_arg` in the very same situation in `Bitv_rel`?

Reply:
Because in the `Bitv_rel` case it *can't* happen (there is a check for the type)! In fact, as noted in my 2nd comment, it is also the case here; the code is right but the comment that says "this case can happen" is wrong.